plugins/secretstores/systemd/README.md

```diff
@@ -0,0 +1,248 @@
+services.
+
+## Credential management
+
+Most steps here are condensed from the [systemd-creds man-page][systemd-creds]
+and are using this command. Please also check that man-page as the options
+or verbs used here might be outdated for the systemd version you are using.
+
+**Please note**: We are using `/etc/telegraf/credentials` as our storage
+location for encrypted credentials throughout the examples below. This is
+because the Telegraf service file expects credentials to be located there
+by default. Furthermore, we assume the secret-store ID to be set to `syscreds`
+in the examples.
+
+### Setup
+
+If you have not done it already, systemd requires a first-time setup of the
+credential system. If you are planning to use the TPM2 chip of your system
+for protecting the credentials, you should first make sure that it is
+available using
+
+```shell
+# sudo systemd-creds has-tpm2
+partial
+-firmware
++driver
++system
++subsystem
+```
+
+The output should look similar to the above.
+
+Now setup the credentials by creating the root key.
+
+```shell
+# sudo systemd-creds setup
+Credential secret file '/var/lib/systemd/credential.secret' is not located on encrypted media, using anyway.
+4096 byte credentials host key set up.
+```
+
+The warning only appears if you are storing the generated key on an unencrypted
+disk which is not recommended. With this, we are all set to create credentials.
+
+### Creating credentials
+
+After setting up the encryption key we can create a new credential using
+
+```shell
+# echo -n "john-doe-jr" | systemd-creds encrypt - /etc/telegraf/credentials/http_user
+```
+
+You should now have a file named `http_user` that contains the encrypted
+username. Please note that systemd credentials are named, so the name
+`http_user` is also stored in the file. The secret-store later provides
+the secret using this name as the secret's key.
+
+You can explicitly name credentials using the `--name` parameter, however,
+in the interest of simplicity, you should follow the filename equals the
+credential/secret's name rule.
+
+We can now add more secrets. To avoid potentially leaking the plain-text
+credentials through command-history or showing it on the screen we use
+
+```shell
+systemd-ask-password -n | systemd-creds encrypt - /etc/telegraf/credentials/http_password
+```
+
+to interactively enter the password.
+
+### Enabling credentials in the service
+
+To actually provide credentials to the Telegraf service, you need to list them
+in the service file. You can use
+
+```shell
+# systemctl edit telegraf
+```
+
+to overwrite parts of the service file. The resulting override should be placed
+in `/etc/systemd/system/telegraf.service.d/override.conf`. The following is an
+example for the content of the file
+
+```text
+[Service]
+LoadCredentialEncrypted=http_user:/etc/telegraf/credentials/http_user
+LoadCredentialEncrypted=http_passwd:/etc/telegraf/credentials/http_passwd
```

Comment:
We should consider documenting the use of `PrivateMounts=` as described in https://systemd.io/CREDENTIALS/:

"It is a good idea to also enable mount namespacing for services that process credentials configured this way. If so, the runtime credential directory of the specific service is not visible to any other service. Use PrivateMounts= as minimal option to enable such namespacing. Note that many other sandboxing settings (e.g. ProtectSystem=, ReadOnlyPaths= and similar) imply PrivateMounts=, hence oftentimes it’s not necessary to set this option explicitly."

Reply:
This *should* be safe since the system mounts should be visible to telegraf but new mounts (like the one for the decrypted files) are not propagated outside of telegraf's private mount namespace, but will need testing. From https://manpages.ubuntu.com/manpages/lunar/en/man5/systemd.exec.5.html
```
PrivateMounts=
           Takes a boolean parameter. If set, the processes of this unit will be run in their own
           private file system (mount) namespace with all mount propagation from the processes
           towards the host's main file system namespace turned off. This means any file system
           mount points established or removed by the unit's processes will be private to them
           and not be visible to the host. However, file system mount points established or
           removed on the host will be propagated to the unit's processes. See
           [mount_namespaces](https://manpages.ubuntu.com/manpages/lunar/en/man7/mount_namespaces.7.html)(7) for details on file system namespaces. Defaults to off.
```

Reply:
I don't think we can do this as it also means that Telegraf cannot read files in the file-system which is required by various plugins...

Reply:
That's the thing; my interpretation of the man page is that the system's private mounts all propagate into the container, but mounts within the container do not propagate into the host. This is also in the man page:
```
           When turned on, this executes three operations for each invoked process: a new
           CLONE_NEWNS namespace is created, after which all existing mounts are remounted to
           MS_SLAVE to disable propagation from the unit's processes to the host (but leaving
           propagation in the opposite direction in effect). Finally, the mounts are remounted
           again to the propagation mode configured with MountFlags=, see below.
```


This is why the flag is meaningful for this feature: without it, the mount is available to others but with it, it is not available to other services (that use PrivateMounts).

I did look into this further and found that:
* when the service is running and `LoadCredentialsEncrypted=...` is used, `$CREDENTIALS_DIRECTORY` is `/run/credentials/telegraf.service` in the service, and this directory has 0500 permissions with `telegraf` as the owner (decrypted files are 0500 with `telegraf` as owner)
* when the service is running and `LoadCredentialsEncrypted=...` is used, the decrypted credentials are in the root-protected `/run/credentials/telegraf.service` directory on the host
* when the service is running and `LoadCredentialsEncrypted=...` is used and `PrivateMounts` IS NOT set, everything in `/run/credentials` on the host is available in `/run/credentials` in the service
* when the service is running and `LoadCredentialsEncrypted=...` is used and `PrivateMounts` IS set, only  `/run/credentials/telegraf.service` on the host is available in `/run/credentials` in the service
* when services that specify `LoadCredentialsEncrypted=...` aren't running, then the decrypted credentials for those services are NOT present in `/run/credentials` on the host (or any running services)

Importantly, this has a few implications:
1. root on the host will always be able to see all the credentials via `/run/credentials` (but also by entering the running the `systemd-creds` command, entering mount namespaces, adjusting unit files, etc). There is no protection against unrestricted root processes on the host
2. with or without `PrivateMounts` being set, the decrypted credentials have DAC permissions set to the `User=` in the systemd unit. Specifically for telegraf, only processes as the `telegraf` user can see the decrypted credentials in `/run/credentials/telegraf.service` on the host or within the service. Other non-root processes cannot see the credentials
3. with `PrivateMounts` NOT SET in a service's unit file that runs as ROOT, the root user can see all currently decrypted credentials for running services (eg, the `foo.service` doesn't specify `User=foo` (or specifies `User=root`), this service can see `/run/credentials/foo.service/*`, `/run/credentials/telegraf.service/*` and any other decrypted credentials for running services)
4. with `PrivateMounts` NOT SET in a service's unit file that runs as NON-ROOT, this non-root user can see all currently decrypted credentials for running services that are accessible to this non-root user
5. with `PrivateMounts` SET in a service's unit file that runs as ROOT, the root user can only see its own decrypted credentials (eg, the `foo.service` doesn't specify `User=foo` (or specifies `User=root`), this service can see `/run/credentials/foo.service/*` but not `/run/credentials/telegraf.service/*` or any other decrypted credentials for running services)
6. with `PrivateMounts` SET in a service's unit file that runs as NON-ROOT, this service can only see `/run/credentials/foo.service/*` but not `/run/credentials/telegraf.service/*` or any other decrypted credentials for running services regardless of if they run as the same user or not)

Practically speaking, this feature does not protect against attacker-controlled root processes or root running services that don't set `PrivateMounts` as these processes can always see the decrypted credentials. `PrivateMounts` adds some protection to the telegraf service against attacker-controlled non-root running services. It adds considerable protection against attacker-controlled root running services when those root running services also specify `PrivateMounts` in their unit files.